internal/pkg/sourceinfer/testdata/src/example.com/tests/core/corecursive.go

```diff
@@ -0,0 +1,45 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"example.com/source"
+)
+
+type (
+	A struct { // want A:"inferred source"
+		source.Source
+		b *B
+	}
+
+	B struct { // want B:"inferred source"
+		a *A
+	}
+)
+
+type (
+	C struct { // want C:"inferred source"
+		source.Source
+		d *D
+	}
+
+	D struct { // want D:"inferred source"
+		e *E
+	}
+
+	E struct { // want E:"inferred source"
+		c *C
+	}
+)
```

Comment:
Your graph approach probably covers this case, but for maximum edge-casing, I'd throw in the following test case:

```golang
type Foo struct {
  neighbors map[string]Foo
}

type Bar struct {
  neighbors map[string]Bar
  s Source
}
```

This sort of recursion forms a tautology, where `Foo` is a source type only if `map[string]Foo` is a source type, only if `Foo` is a source.  That ambiguity should default to "not a source", but should be broken by ties like in `Bar`.

I guess you could probably get the same result with just a `type A struct { *A }`.  So some directly-recursive tests might be good.  Negative co-recursive cases would also be good.

Reply:
These are very good cases. I've added negative corecursive cases, as well as positive and negative self-recursive cases.